pkgs/build-support/singularity-tools/default.nix

```diff
@@ -108,14 +116,14 @@ rec {
 
             # Build /bin and copy across closure
             mkdir -p bin ./${builtins.storeDir}
-            for f in $(cat $layerClosure) ; do
-              cp -ar $f ./$f
-            done
+            while IFS= read -r f; do
+              cp -r "$f" "./$f"
```

Comment:
I wonder why `-a` is not necessary. I'm seeing that if I do

```nix
$ touch a
$ chmod +x a
$ cp a b
$ stat b
```

...`b`s still has the +x bit, but I don't know why

Reply:
> I'm seeing that if I do
> 
> ```nix
> $ touch a
> $ chmod +x a
> $ cp a b
> $ stat b
> ```
> 
> ...`b`s still has the +x bit, but I don't know why

For `cp`, `-a` (`--archive`) implies `--preserve all`, which means preserving all *attributes*, including `mode`, `ownership`, `timestamps`, `links` (hard links), `context` (SELinux or SMACK security context), `xattr`. `cp` preserves the file mode by default but not the file ownership.

It requires root privileges to create a file owned by `root:root`, which is why it requires root privileges to `cp -ra` from a store object.

> I wonder why `-a` is not necessary.

The initial purpose of preserving ownership might be to please Nix when using Nix inside the container. However, the container Nix store is inherently broken from Nix's perspective, as it doesn't include the database.

Cc: @jbedo, the original author of `singularity-tools`.

Reply:
> It requires root privileges to create a file owned by root:root

It doesn't? In particular, this works:

```
with import <nixpkgs> { };

runCommand "abc" { } ''
  cp -ar ${hello}/ $out/
''
```

Reply:
> > It requires root privileges to create a file owned by root:root
> 
> It doesn't? In particular, this works:
> 
> ```
> with import <nixpkgs> { };
> 
> runCommand "abc" { } ''
>   cp -ar ${hello}/ $out/
> ''
> ```

Oops! I remember that it wasn't when I first worked on it.